lib/SpecializeImageTypes.cpp

```diff
@@ -139,47 +198,46 @@ Type *SpecializeImageTypesPass::RemapUse(Value *value, unsigned operand_no) {
         break;
       }
 
-      // Read only images are translated as sampled images.
-      const auto pos = name.find("_wo_t");
-      if (!IsStorageImageType(imageTy)) {
+      const auto wo_pos = name.find("_wo");
+      const auto ro_pos = name.find("_ro");
+      if (ro_pos != std::string::npos) {
         name += ".sampled";
-      } else if (clspv::Option::Language() >=
-                     clspv::Option::SourceLanguage::OpenCL_C_20 &&
-                 pos != std::string::npos) {
+      } else if (Option::Language() >= Option::SourceLanguage::OpenCL_C_20 &&
+                 wo_pos != std::string::npos) {
         // In OpenCL 2.0 (or later), treat write_only images as read_write
         // images. This prevents the compiler from generating duplicate image
         // types (invalid SPIR-V).
-        name = name.substr(0, pos) + "_rw_t" + name.substr(pos + 5);
+        name = name.substr(0, wo_pos) + "_rw" + name.substr(wo_pos + 3);
       }
 
-      StructType *new_struct =
-          StructType::getTypeByName(call->getContext(), name);
-      if (!new_struct) {
-        new_struct = StructType::create(call->getContext(), name);
+      struct_ty = StructType::getTypeByName(call->getContext(), name);
+      if (!struct_ty) {
+        struct_ty = StructType::create(call->getContext(), name);
       }
-      PointerType *new_pointer =
-          PointerType::get(new_struct, imageTy->getPointerAddressSpace());
-      return new_pointer;
+      return std::make_pair(ResultType::kSpecialized, struct_ty);
     }
+    // Non-specializable cases: queries.
     default:
-      if (!called->isDeclaration()) {
-        for (auto &U : called->getArg(operand_no)->uses()) {
-          if (auto new_type = RemapUse(U.getUser(), U.getOperandNo())) {
-            return new_type;
-          }
-        }
-      }
+      return std::make_pair(ResultType::kNotSpecialized, struct_ty);
       break;
     }
-  } else if (IsImageType(value->getType())) {
+  } else if (value->getType()->isPointerTy()) {
```

Comment:
nit: When does this case occur?

Is it analogous to OpCopyObject of what looks like a pointer argument?  But I'm not sure if that happens in the LLVM IR.  A comment here would be helpful.

Reply:
I know it happens for address space casts. I'll add a comment.